### PR TITLE
feat: make project delivery management report basic provider

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
project delivery management report basic provider
make prometheus collector for iteration
at present, the provider only collector iteration basic metrics for project

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=453638&iterationID=1841&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |   Support make project delivery management report basic provider           |
| 🇨🇳 中文    |    实现项目交付报表基础指标采集的provider          |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: cmd/erda-server/bootstrap.yaml [hunk 1]
```diff
@@ -23,6 +23,9 @@ health:
 component-protocol:
 openapi-component-protocol:
 
+prometheus:
+  router_label_enable: false
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:27]: What's `router_label_enable`, how about set it to `true`.

File: cmd/erda-server/bootstrap.yaml [hunk 2]
```diff
@@ -279,6 +282,7 @@ uc-adaptor:
 dop: { }
 erda.core.dop.taskerror: { }
 erda.dop.qa.unittest: { }
+project-report: { }
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:285]: We should unify the term. I think `project-report` may contains other metrics more than issue-management.
So we should use `project-management-report` here.

File: internal/apps/dop/providers/project_report/cache.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	values map[string]*cacheEntry
+	freqs  *list.List
+	len    int
+}
+
+type orgCache struct {
+	*Cache
+}
+
+type listEntry struct {
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/cache.go:41]: listCacheEntry

File: internal/apps/dop/providers/project_report/cache.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+	}
+}
+
+func (c *Cache) remEntry(place *list.Element, entry *cacheEntry) {
+	entries := place.Value.(*listEntry).entries
+	delete(entries, entry)
+	if len(entries) == 0 {
+		c.freqs.Remove(place)
+	}
+}
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/cache.go:233]: How about using a cache library directly? Such as https://github.com/patrickmn/go-cache

File: internal/apps/dop/providers/project_report/calculate.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	}, true)
+	if err != nil {
+		p.Log.Errorf("failed to get all orgs, err: %v", err)
+		return err
+	}
+	for i := range orgs {
+		p.orgSet.Set(orgs[i].ID, orgs[i])
+	}
+
+	projects, err := p.bdl.GetAllProjects()
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/calculate.go:40]: Bundle API may contains many useless information, please check this API and maybe you should use dbclient directly.

Comment (sfwn) [reply internal/apps/dop/providers/project_report/calculate.go:40]: check other bundle APIs

File: internal/apps/dop/providers/project_report/calculate.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	for _, projectDto := range projects {
+		orgDto := p.orgSet.Get(projectDto.OrgID)
+		if orgDto == nil {
+			p.Log.Errorf("failed to find org for project: %s, projectID: %d, orgID: %d",
+				projectDto.Name, projectDto.ID, projectDto.OrgID)
+			continue
+		}
+		iterations, err := p.bdl.ListProjectIterations(apistructs.IterationPagingRequest{
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/calculate.go:56]: why not ListAllProjectIterations at once?

File: internal/apps/dop/providers/project_report/provider.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/erda-project/erda/internal/tools/pipeline/providers/reconciler/rutil"
+	"github.com/erda-project/erda/pkg/jsonstore"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+type config struct {
+	ProjectReportMetricDuration  time.Duration `file:"project_report_metric_duration" env:"PROJECT_REPORT_METRIC_DURATION" default:"2h"`
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/provider.go:40]: What is the duration for?

File: internal/apps/dop/providers/project_report/metrics.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	IterationCompleteSchedule float64
+}
+
+// IsValid returns true if the IterationMetricFields is valid.
+// we need to ensure that IterationMetricFields is the data of the day to avoid double calculation
+func (i *IterationMetricFields) IsValid() bool {
+	return i.CalculatedAt.Day() == time.Now().Day()
+}
+
+var invalidNameCharRE = regexp.MustCompile(`[^a-zA-Z0-9_]`)
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/metrics.go:145]: Also apply this rule when query.

File: internal/apps/dop/providers/project_report/metrics.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	orgpb "github.com/erda-project/erda-proto-go/core/org/pb"
+	"github.com/erda-project/erda/apistructs"
+)
+
+const (
+	metricGroupName = "project_delivery_report"
```
Comment (sfwn) [inline internal/apps/dop/providers/project_report/metrics.go:29]: `management`

**Review by sfwn — CHANGES_REQUESTED** [review]

File: cmd/erda-server/bootstrap.yaml (hunk 1 above)
Comment (chengjoey) [reply cmd/erda-server/bootstrap.yaml:27]: the `router_label_enable` default value is true, it find svc named `http-router@admin`

we should open the metrics in 9095 `http-server`, os set it to false

File: cmd/erda-server/bootstrap.yaml (hunk 2 above)
Comment (chengjoey) [reply cmd/erda-server/bootstrap.yaml:285]: adjusted

File: internal/apps/dop/providers/project_report/cache.go (hunk 4 above)
Comment (chengjoey) [reply internal/apps/dop/providers/project_report/cache.go:233]: good idea, i adjusted using the go-cache library directly

File: internal/apps/dop/providers/project_report/calculate.go (hunk 5 above)
Comment (chengjoey) [reply internal/apps/dop/providers/project_report/calculate.go:40]: adjusted, use dbclient and get all projects directly

File: internal/apps/dop/providers/project_report/calculate.go (hunk 6 above)
Comment (chengjoey) [reply internal/apps/dop/providers/project_report/calculate.go:56]: adjusted

File: internal/apps/dop/providers/project_report/provider.go (hunk 7 above)
Comment (chengjoey) [reply internal/apps/dop/providers/project_report/provider.go:40]: removed

File: apistructs/issue.go [hunk 10]
```diff
@@ -1159,8 +1159,9 @@ type IssueTestCaseRelationsListRequest struct {
 
 // IssueManHourSumRequest 事件下所有的任务总和请求
 type IssuesStageRequest struct {
-	StatisticRange string `json:"statisticRange"` //事件类型 项目/迭代
-	RangeID        int64  `json:"rangeId"`        //项目id/迭代id
+	StatisticRange string  `json:"statisticRange"` //事件类型 项目/迭代
+	RangeID        int64   `json:"rangeId"`        //项目id/迭代id
+	StateIDS       []int64 `json:"stateIds"`       // state id list
```
Comment (sfwn) [inline apistructs/issue.go:1164]: StateIDs.

Adjust in next pr please.

**Review by sfwn — APPROVED** [review]

**Pull request merged**